lib/src/session.dart

```diff
@@ -0,0 +1,182 @@
+  Stream<SyncProgress> getProgressStream(ProgressDirection direction, ProgressMode mode) {
+    final controller = SessionProgressNotificationsController(this, direction, mode);
+    return controller.createStream();
+  }
+}
+
+/// The current state of a [Session] object
+enum SessionState {
+  /// The session is connected to the MongoDB Realm server and is actively transferring data.
+  active,
+
+  /// The session is not currently communicating with the server.
+  inactive,
+}
+
+/// The current connection state of a [Session] object
+enum ConnectionState {
+  /// The session is disconnected from the MongoDB Realm server.
+  disconnected,
+
+  /// The session is connecting to the MongoDB Realm server.
+  connecting,
+
+  /// The session is connected to the MongoDB Realm server.
+  connected,
+}
+
+/// The transfer direction (upload or download) tracked by a given progress notification subscription.
+enum ProgressDirection {
+  /// Monitors upload progress.
+  upload,
+
+  /// Monitors download progress.
+  download
+}
+
+/// The desired behavior of a progress notification subscription.
+enum ProgressMode {
+  /// The callback will be called forever, or until it is unregistered by closing the `Stream<SyncProgress>`.
+  /// Notifications will always report the latest number of transferred bytes, and the most up-to-date number of
+  /// total transferable bytes.
+  reportIndefinitely,
+
+  /// The callback will, upon registration, store the total number of bytes to be transferred. When invoked, it will
+  /// always report the most up-to-date number of transferable bytes out of that original number of transferable bytes.
+  /// When the number of transferred bytes reaches or exceeds the number of transferable bytes, the callback will
+  /// be unregistered.
+  forCurrentlyOutstandingWork,
+}
+
+/// A type containing information about the progress state at a given instant.
+class SyncProgress {
+  /// The number of bytes that have been transferred since subscribing for progress notifications.
+  final int transferredBytes;
+
+  /// The total number of bytes that have to be transferred since subscribing for progress notifications.
+  /// The difference between that number and [transferredBytes] gives you the number of bytes not yet
+  /// transferred. If the difference is 0, then all changes at the instant the callback fires have been
+  /// successfully transferred.
+  final int transferableBytes;
+
+  SyncProgress._(this.transferredBytes, this.transferableBytes);
+}
+
+extension SessionInternal on Session {
+  static Session create(SessionHandle handle, Scheduler scheduler) => Session._(handle, scheduler);
+
+  SessionHandle get handle => _handle;
+
+  Scheduler get scheduler => _scheduler;
+}
+
+/// @nodoc
+class SessionProgressNotificationsController {
+  final Session _session;
+  final ProgressDirection _direction;
+  final ProgressMode _mode;
+
+  int? _token;
```

Comment:
It is maybe not a big problem, and we are already leaking the `SessionProgressNotificationsController`, if user doesn't cancel their subscriptions, but having the token as just a pointer to a int won't ensure that it is probably unsubscribed when torn down. We should perhaps wrap it with a finalizable wrapper.

Anyway, that can come later, if we decide to get rid of the persistent handle, or fix the leak another way.

Reply:
We can do use a managed finalizer once we upgrade to 2.17.